### PR TITLE
wip/v2 api proposal

### PR DESCRIPTION
**Pull request by bcgodev**
With the finishing deployment of `tv-cluster`, `major-tom-go v2` aims to support the operations of `tv-cluster`.

I would like to know some feedback or concerns for the design.

I also have a concern addressing the usage of service name, which could be quite long in the case of `openwarehouse`. I am not sure if there's a better way to specify the target. I'd like to know any suggestion. Thanks!

# major-tom-go v2

`major-tom-go` is a bot written in golang to help change the configurations on kubernetes easily via GitOps and will play a role in the CI pipeline.

The `kubernetes-configs` repository will be the sole resource of GitOps.

## Proposal

The following is the proposal of API design for `major-tom-go` `v2`, and `v2` will be a breaking change.

`major-tom-go v2` will be referred as `major tom` below.
## Operations

`major tom` should support the following operations:

1. Change the image tag to be deployed
2. Change the autoscaling configurations for deployments
3. Change the current pod numbers
4. List the changes in history
5. List the supported resource
6. Show the current deployment info

These operation should be supported services in all stages of all projects in the end, but the **initial implementation** will only support the `tv` project and change of image tag.

All projects, including `readr`, `weekly`, and `tv` will be supported at the end.

## API Design Proposal

I would separate the operation to four categories:

1. Image change
2. Scaling
3. List
4. Info

According to the design of `kubernetes-configs`, the image tag will be defined in different level of folders from `prod` to `staging` and `dev`.

### Image change

Thus two different api are designed for them. Reasons will be given after the following section.

1. `release`: means release a image tag
   1. `release {repo} project={project} image-tag={image tag}` for the likes of `openwarehouse`, eg., `release openwarehouse project=tv image-tag=prod_81ab7ac`
   2.  `release {repo} image-tag={image tag}` for the likes of `mirror-tv-nuxt`, eg., `release mirror-tv-nuxt image-tag=prod_399440e`

It will look for the `kustomize.yaml` in the project folder, if specified, of `prod` by default.

2. `deploy`: can deploy a image to staing or dev environments for all projects
   1. `deploy {repo} env={stage} image-tag={image tag}`, eg., `deploy openwarehouse env=dev mage-tag=dev_1bac23`

I choose two separate commands to change image tag in different environments because

- Their structure in `kubernetes-configs` are different.
- `deploy` will deploy image to all projects but `release` will only deploy the image to the specific project. They are essentially two different operations.
- `release` emphasize the importance of the command and help communication.
- Combining them would make the command implementation cumbersome.

### Scaling

`scale`: 

`scale {service} env={stage} conf1=value1 conf2=value2 ...` for the likes of `openwarehouse`

For example: `scale openwarehouse-tv-gql-external env=prod maxReplicas=1`

or 

`scale mirror-tv-nuxt env=staging maxReplicas=3 minReplicas=1`

Because we established a [naming and folder structure convention](https://github.com/mirror-media/kubernetes-configs/commit/f0fc6c2ca18e1c3a8fb6df50a4882121d5b67548) for `kubernetes-configs`, `major tom` doesn't need the repo name to find the desired `kustomize.yaml`. With a service name merely, `major tom` can identify the repo with confidence.

#### The supported conf are

- maxReplicas
- minReplicas

and they are not required to be both supplied in the same command.

Their values should be a positive number and `major tom` will validate them.

`major tom` will look for `hpa.yaml` in the service folder or stage folder depending on the structure of the repo.

If `major tom` doesn't find a `hpa.yaml` at the desired location, it will respond with `auto scaling is not support for the service`.

If we specify a replica number in the `Deployment` resource, then it will always try deploy the number each time. It may have unwanted side-effect, so leaving it to autoscaling would be a better idea.

To change the current replica, one could use `scale` to change the scale in or scale out immediately.

### List

`list`:

`list project={project} env={stage}` should provide a list of supported service name and repo name in the specified stage.

For example, `list project=tv env=staging`

Response should be a list of `{repo}/{service}` or `{repo}` like
```
1. `openwarehouse/openwarehouse-tv-cms`
2. `openwarehouse/openwarehouse-tv-gql-external`
3. `yt-relay/yt-relay-tv-yt-relay`
4. `mirror-tv-nuxt`
```

`list {service} env={stage}` or `list {repo} env={stage}` should provide the latest eleven changes in history.

### Info

`info`:

`info {service} env={stage}` or `info {repo} env={stage}` to report current status on kubernetes.

For example, `info openwarehouse-tv-gql-internal env=prod` or `info mirror-tv-nuxt env=dev`

Status includes `image tag`, `scaling configs`, and `current replica`.

### CODE REVIEW

**Comment by hcchien** [comment]
"deploy will deploy image to all projects but release will only deploy the image to the specific project. They are essentially two different operations."  The 'project' here means 'tv', 'readr' and/or 'weekly', right?

**Comment by hcchien** [comment]
deploy {repo} {stage} {image tag}

But the major_tom will deploy the production only?

**Comment by bcgodev** [comment]
> "deploy will deploy image to all projects but release will only deploy the image to the specific project. They are essentially two different operations." The 'project' here means 'tv', 'readr' and/or 'weekly', right?

Yes. That's right.

> deploy {repo} {stage} {image tag}
>
> But the major_tom will deploy the production only?

There are two commands:

`deploy` will handle the `dev` and staging` environments.
`release` will be responsible of `prod` only.

**Comment by hcchien** [comment]
> @hcchien
> 
> > "deploy will deploy image to all projects but release will only deploy the image to the specific project. They are essentially two different operations." The 'project' here means 'tv', 'readr' and/or 'weekly', right?
> 
> Yes. That's right.
> 
> I added a projects list to the PR to clarify it.
> 
> > deploy {repo} {stage} {image tag}
> > But the major_tom will deploy the production only?
> 
> There are two commands:
> 
> `deploy` will handle the `dev` and `staging` environments.
> 
> `release` will be responsible of `prod` only.

ok. so 'deploy' will be done by major_tom automatically after it got the build success message in slack, and we just handle the 'release'.  Is my understanding correct?

**Comment by bcgodev** [comment]
> ok. so 'deploy' will be done by major_tom automatically after it got the build success message in slack, and we just handle the 'release'. Is my understanding correct?

Yes!

**Comment by bcgodev** [comment]
> > @hcchien
> > > "deploy will deploy image to all projects but release will only deploy the image to the specific project. They are essentially two different operations." The 'project' here means 'tv', 'readr' and/or 'weekly', right?
> > 
> > 
> > Yes. That's right.
> > I added a projects list to the PR to clarify it.
> > > deploy {repo} {stage} {image tag}
> > > But the major_tom will deploy the production only?
> > 
> > 
> > There are two commands:
> > `deploy` will handle the `dev` and `staging` environments.
> > `release` will be responsible of `prod` only.
> 
> ok. so 'deploy' will be done by major_tom automatically after it got the build success message in slack, and we just handle the 'release'. Is my understanding correct?

Yes, but the build message must issue a `deploy` command for major-tom to acknowledge.

**Comment by bcgodev** [comment]
> Yes, but the build message must issue a `deploy` command for major-tom to acknowledge.

This is part of the CI/CD pipeline, instead of `major-tom`'s concern. As long as `major tom` receives a `deploy` command regardless of the user, it will change the the image tag in `kubernetes-configs`, to trigger CD.

**Comment by hcchien** [comment]
> > > @hcchien
> > > > "deploy will deploy image to all projects but release will only deploy the image to the specific project. They are essentially two different operations." The 'project' here means 'tv', 'readr' and/or 'weekly', right?
> > > 
> > > 
> > > Yes. That's right.
> > > I added a projects list to the PR to clarify it.
> > > > deploy {repo} {stage} {image tag}
> > > > But the major_tom will deploy the production only?
> > > 
> > > 
> > > There are two commands:
> > > `deploy` will handle the `dev` and `staging` environments.
> > > `release` will be responsible of `prod` only.
> > 
> > 
> > ok. so 'deploy' will be done by major_tom automatically after it got the build success message in slack, and we just handle the 'release'. Is my understanding correct?
> 
> Yes, but the build message must issue a `deploy` command for major-tom to acknowledge.

ok. it would be fine to me then.

**Comment by nickhsine** [comment]
Could you provide command examples here?
It will make the document more easy to understand.

**Comment by nickhsine** [comment]
Could we 
1. sort command args in the same order? Right now, we have 
```
scale {service} {stage} {project} conf1:value1 conf2:value2
list {project} {stage}
info {project} {stage} {service}
list {project} {stage} {repo}
```
different orders.

2. for PM or other developers needing MajorTom, do they need to know the `{repo}` name like
```
1. `openwarehouse/openwarehouse-tv-cms`
2. `openwarehouse/openwarehouse-tv-gql-external`
3. `yt-relay/yt-relay-tv-yt-relay`
```

It's kind of confusing if they don't read `kubernetes-configs` before.

Would it be user-friendly if users just need to know
1. stage (dev, staging, prod)
2. project (tv, readr, weekly)
3. serviceName (mirror-tv-nuxt, openwarehouse-tv-cms, openwarehouse-tv-gql-internal) ... etc

**Comment by bcgodev** [comment]
> Could you provide command examples here?
> It will make the document more easy to understand.

I added some examples to each command, and remove `{repo}` from info and scale

**Comment by bcgodev** [comment]
> Would it be user-friendly if users just need to know
> 
>     1. stage (dev, staging, prod)
> 
>     2. project (tv, readr, weekly)
> 
>     3. serviceName (mirror-tv-nuxt, openwarehouse-tv-cms, openwarehouse-tv-gql-internal) ... etc

I have thought about that too. It would indeed be more simpler, but `deploy openwarehouse-tv-cms` would  update all service including `openwarehouse-tv-gql-internal` and `openwarehouse-tv-cms`, that would be confusing instead.

Ideally PM will only use `release` command to deploy image to production, so PM will only use repo name and project name in this case, which would be provided by `list`.

I agree PM should have to understand the nature of `kubernetes-configs`. I would love to discuss any possible interface to make the operation more easier.

**Comment by bcgodev** [comment]
> Could we
> 
>     1. sort command args in the same order? Right now, we have
> 
> 
> ```
> scale {service} {stage} {project} conf1:value1 conf2:value2
> list {project} {stage}
> info {project} {stage} {service}
> list {project} {stage} {repo}
> ```
> 
> different orders.
> 

That's a great suggestion! I would review the command and make such changes later. Thanks.

**Comment by nickhsine** [comment]
>> 3. serviceName (mirror-tv-nuxt, openwarehouse-tv-cms, openwarehouse-tv-gql-internal) ... etc

> I have thought about that too. It would indeed be more simpler, but deploy openwarehouse-tv-cms would update all service including openwarehouse-tv-gql-internal and openwarehouse-tv-cms, that would be confusing instead.

For me, it is wonderful if I could do the following commands
Take `openwarehouse`, the complicated one, for example:
```
// deploy `openwarehouse-tv-cms`, `openwarehouse-tv-gql-internal`, `openwarehouse-tv-gql-external` with new image to dev environment.
`major-tom-2 deploy openwarehouse --env=dev --project=tv --image-tag=dev_1bac23`

// deploy `openwarehouse-readr-cms`, `openwarehouse-tv-gql-internal`, `openwarehouse-tv-gql-external` with new image to staging environment.
`major-tom-2 deploy openwarehouse --env=staging --project=readr --image-tag=staging_1bac23` 

// deploy `openwarehouse-readr-cms`, `openwarehouse-tv-gql-internal`, `openwarehouse-tv-gql-external` with new image to staging environment.
`major-tom-2 deploy openwarehouse --env=prod --project=weekly --image-tag=staging_1bac23` 

// which is equivalent to the above one
`major-tom-2 release openwarehouse --project=tv --image=tag=staging_1bac23`
```

**Comment by bcgodev** [comment]
@nickhsine Great suggestion! I like the usage of env and project as arguments and the equal sign.

In the current philosophy, I'd like to use a slashcommand and separate each arguments by space only.

// deploy `openwarehouse-tv-cms`, `openwarehouse-tv-gql-internal`, `openwarehouse-tv-gql-external` with new image to dev environment.
`major-tom-2 deploy openwarehouse --env=dev --project=tv --image-tag=dev_1bac23`

// deploy `openwarehouse-readr-cms`, `openwarehouse-readr-gql-internal`, `openwarehouse-readr-gql-external` with new image to staging environment for all project
`/major-tom deploy openwarehouse env=staging project=readr image-tag=staging_1bac23` 

// deploy `openwarehouse-weekly-cms`, `openwarehouse-weekly-gql-internal`, `openwarehouse-weekly-gql-external` with new image to prod environment for project weekly
`/major-tom release openwarehouse project=weekly image-tag=staging_1bac23`

I would also remove the support of `prod` for `deploy` to simplify the requirements of arguments.

Will it do for you?

**Comment by bcgodev** [comment]
@nickhsine BTW, I had missed a word `not` in previous comment
```I agree PM should not have to understand the nature of kubernetes-configs. I would love to discuss any possible interface to make the operation more easier.```

**Comment by nickhsine** [comment]
> I would also remove the support of `prod` for `deploy` to simplify the requirements of arguments.
> 
> Will it do for you?

That will be perfect for me 👍 .

Just one concern I think of is:
in our plan, we release _dev/staging_ openwarehouse for three projects by default.
Image tag will be modified in `openwarehouse/overlays/(dev|staging)/base`.

However, if we want to provide single project deployment, such as only deploy dev openwarehouse for tv project,
and then we might need another image tag which could be modified in `openwarehouse/overlays/dev/overlays/tv/base` to overwrite the image tag in `openwarehouse/overlays/dev/base`. 

Is it workable for you?
Will it make kubernetes-configs harder to understand due to multiple image tags?

**Comment by bcgodev** [comment]
> However, if we want to provide single project deployment, such as only deploy dev openwarehouse for tv project,
> and then we might need another image tag which could be modified in `openwarehouse/overlays/dev/overlays/tv/base` to overwrite the image tag in `openwarehouse/overlays/dev/base`.

> Will it make kubernetes-configs harder to understand due to multiple image tags?
It did happen to be in a [PR](https://github.com/mirror-media/kubernetes-configs/pull/40) of `kubernetes-configs`, but ,as long as it's documented, the readability will be fine to me.

As for case of a specific deployment for single project in `dev` and `staging` stage, I would suggest to plan in the future instead of now. The future problem could be solved once it arrives.

Although overwriting image tag is easy to understand in the perspective of `kustomize`, but it seems more complicated in the perspective of current CI plan. I'd hesitate to design to fix the problem right now.

**Comment by bcgodev** [comment]
@nickhsine I updated the command format and examples. Please help review them. Thanks.

File: v2/README.md [hunk 1]
```diff
@@ -40,26 +42,32 @@ According to the design of `kubernetes-configs`, the image tag will be defined i
 Thus two different api are designed for them. Reasons will be given after the following section.
 
 1. `release`: means release a image tag
-   1. `release {repo} {project} {image tag}` for the likes of `openwarehouse`
-   2.  `release {repo} {image tag}` for the likes of `mirror-tv-nuxt`
+   1. `release {repo} project={project} image-tag={image tag}` for the likes of `openwarehouse`, eg., `release openwarehouse tv prod_81ab7ac`
+   2.  `release {repo} image-tag={image tag}` for the likes of `mirror-tv-nuxt`, eg., `release mirror-tv-nuxt prod_399440e`
```
Comment (nickhsine) [inline v2/README.md:46]: `release mirror-tv-nuxt prod_399440e` -> `release mirror-tv-nuxt image-tag=prod_399440e` ?

File: v2/README.md [hunk 2]
```diff
@@ -40,26 +42,32 @@ According to the design of `kubernetes-configs`, the image tag will be defined i
 Thus two different api are designed for them. Reasons will be given after the following section.
 
 1. `release`: means release a image tag
-   1. `release {repo} {project} {image tag}` for the likes of `openwarehouse`
-   2.  `release {repo} {image tag}` for the likes of `mirror-tv-nuxt`
+   1. `release {repo} project={project} image-tag={image tag}` for the likes of `openwarehouse`, eg., `release openwarehouse tv prod_81ab7ac`
```
Comment (nickhsine) [inline v2/README.md:45]: `release openwarehouse tv prod_81ab7ac` -> `release openwarehouse project=tv image-tag=prod_81ab7ac`?

File: v2/README.md [hunk 3]
```diff
@@ -40,26 +42,32 @@ According to the design of `kubernetes-configs`, the image tag will be defined i
...
+   2.  `release {repo} image-tag={image tag}` for the likes of `mirror-tv-nuxt`, eg., `release mirror-tv-nuxt prod_399440e`
 
 It will look for the `kustomize.yaml` in the project folder, if specified, of `prod` by default.
 
-1. `deploy`: can deploy a image to staing or dev environments for all projects
-   1. `deploy {repo} {stage} {image tag}`
+2. `deploy`: can deploy a image to staing or dev environments for all projects
+   1. `deploy {repo} env={stage} image-tag={image tag}`, eg., `deploy openwarehouse dev dev_1bac23`
```
Comment (nickhsine) [inline v2/README.md:51]: `deploy openwarehouse dev dev_1bac23` -> `deploy openwarehouse env=dev image-tag=dev_1bac23` ?

File: v2/README.md [hunk 4]
```diff
@@ -94,12 +104,14 @@ Response should be a list of `{repo}/{service}` or `{repo}` like
 4. `mirror-tv-nuxt`
 ```
 
-`list {project} {stage} {service}` or `list {stage} {repo}` should provide the latest eleven changes in history.
+`list {service} project={project} env={stage} ` or `list {stage} {repo}` should provide the latest eleven changes in history.
```
Comment (nickhsine) [inline v2/README.md:107]: `list {stage} {repo}` -> `list {repo} env={stage}` ?

Comment (bcgodev) [reply v2/README.md:107]: Yes. And `list {service} project={project} env={stage}` should be `list {service}  env={stage}`. The project argument is redundant.

File: v2/README.md (hunk 1 above)
Comment (bcgodev) [reply v2/README.md:46]: Updated in f4955e5. Please check the update. 😄

File: v2/README.md (hunk 2 above)
Comment (bcgodev) [reply v2/README.md:45]: Updated in f4955e5. Please check the update. 😄

File: v2/README.md (hunk 3 above)
Comment (bcgodev) [reply v2/README.md:51]: Updated in f4955e5. Please check the update. 😄

File: v2/README.md (hunk 4 above)
Comment (bcgodev) [reply v2/README.md:107]: Updated in f4955e5. Please check the update. 😄

**Comment by nickhsine** [comment]
LGTM 👍 💯 🥇 
I am looking forward to _MajorTom_ !

**Comment by bcgodev** [comment]
Thanks for everyone's efforts. Now the PR has concluded. I'm closing the PR and leaving it as a documented record.

**Comment by bcgodev** [comment]
Fell free to reopen the PR if conversation is needed.

**Pull request closed without merging**